### ISSUE TITLE
td3_implementation analysis

### ISSUE THREAD

**Issue by dion-jy** [CUN-bjy/gym-td3-keras#10]
the first TD3 implementation do not work well..

so.. have to analysis each part of the differences from ddpg

**Comment by dion-jy** [comment]
**[TEST 1]**
- rebase to ddpg
- add **target policy smoothing** term.
- test on **RoboschoolInvertedPendulum-v1.**
- batch_size -> **64**, hidden_layers -> **24, 16**(for each actor and critic)
- lr-> **1e-4,1e-3**, tau->**1e-3,1e-3**(for each actor and critic)
- _it works_.(cause of exploration problem, I tested 5 times more.. success rate is under 30%.)
:heavy_check_mark:  

**[TEST 2]**
- basically, tested on InvertedPendulum, almost same to above parameters and added target policy smoothing.
- change parameters, lr->**3e-4,3e-4**, tau->**5e-3,5e-3** (same to original td3 code)
- _it works_.(also have exploration problem, success rate is not high.)
:heavy_check_mark: 

**[TEST 3]**
- add **delayed policy update** term.
- delayed update peoriod -> 2
- _it works_.(have exploration problem, success rate is too low, and feel like delayed to learn a good policy)
:heavy_check_mark:

**Comment by dion-jy** [comment]
**[TEST 4]**
- rebase to ddpg
- add **target policy smoothing** term.
- test on RoboschoolInvertedPendulum-v1.
- batch_size -> **64**, hidden_layers -> **24, 16**(for each actor and critic)
- lr->**3e-4,3e-4**, tau->**5e-3,5e-3** (for each actor and critic)
- (reset policy update interval to 1)
- add **double cliped Q update** term
- but **only use Q1 for target update**
-  _it works_.
:heavy_check_mark: 

**[TEST 5]**
- same to above experiment.
- **use Q1,Q2** for target update
- it doesn't work well..

![Screenshot from 2021-01-28 19-19-59](https://user-images.githubusercontent.com/26274945/106123897-d8345480-619d-11eb-9f96-99ba5fd71178.png)

![Screenshot from 2021-01-28 19-19-07](https://user-images.githubusercontent.com/26274945/106123818-bfc43a00-619d-11eb-837d-7f8faf93ee69.png)

**[TEST 6]**
- test on RoboschoolInvertedPendulum-v1.
- TD3 set.
- add **target policy smoothing** term.
- add **delayed policy update** term. update_interval -> 2
- add **double cliped Q update** term
- batch_size -> **64**, hidden_layers -> **24, 16**(for each actor and critic)
- lr->**3e-4,3e-4**, tau->**5e-3,5e-3** (for each actor and critic)
- also doesn't work well..(take long time to learn the policy and getting worse)

### **:eyes: I think the my implementation of `double cliped q update` has some problems.**

**Comment by dion-jy** [comment]
**[TEST 7]**
- test on RoboschoolInvertedPendulum-v1.
- add **double cliped Q update** term **(only)**
- batch_size -> 64, hidden_layers -> 24, 16(for each actor and critic)
- lr->3e-4,3e-4, tau->5e-3,5e-3 (for each actor and critic)
- yes.. doesn't work.. **Catastrophic forgetting problem** happen, even in one task.

**[TEST 8]**
- same condition to above.
- changed some codes that make i**ndependent optimizer for critics**
![Screenshot from 2021-02-06 21-47-36](https://user-images.githubusercontent.com/26274945/107119550-b76dac80-68cb-11eb-8ad6-a08ec56c687b.png)
![Screenshot from 2021-02-06 21-47-41](https://user-images.githubusercontent.com/26274945/107119552-b9d00680-68cb-11eb-96b2-565065c29301.png)
- I think that **is working**, but very slow compared to simple ddpg.
- by double cliped q update overestimation problem is fixed, but that make policy update so slow.
-  should try integrated system and than should change some parameters..

**Comment by CUN-bjy** [comment]
trying independent optimizer for critics

**Comment by dion-jy** [comment]
**[TEST 9]**
- integrate that all.. -> doesn't work..
![reward](https://user-images.githubusercontent.com/26274945/107132806-7e622600-6925-11eb-9b04-74f6ccecf019.png)
![critic_loss](https://user-images.githubusercontent.com/26274945/107132804-7b673580-6925-11eb-86a0-811a49d2934f.png)

**Comment by dion-jy** [comment]
**[TEST 10]**
- **initial random policy** added for exploration
- and there's some mistake..

**before**
```python
		a = agent.make_action(obs,t)
		action = np.argmax(a) if is_discrete else a

		# do step on gym at t-step
		new_obs, reward, done, info = env.step(action) 

		# store the results to buffer	
		agent.memorize(obs, a, reward, done, new_obs) 
                # should've memorize action w/ noise!!
```
**after**
```python
		a = agent.make_action(obs,t)
		action = np.argmax(a) if is_discrete else a

		# do step on gym at t-step
		new_obs, reward, done, info = env.step(action) 

		# store the results to buffer	
		agent.memorize(obs, action, reward, done, new_obs)
```

but, consequently, doesn't work..

**Comment by dion-jy** [comment]
**[TEST 11]**
- use OU noise process for off-policy exploration strategy 
![reward](https://user-images.githubusercontent.com/26274945/109091051-6842cb00-7757-11eb-862a-16a6f39b2667.png)
![critic_loss](https://user-images.githubusercontent.com/26274945/109091047-65e07100-7757-11eb-8009-d8f37209fcde.png)

also, doesn't work.

**Comment by dion-jy** [comment]
[TEST 12]
- use OU noise process for off-policy exploration strategy
- without BatchNormalization, Weight Regularization, Groot Initializer on actor and critic.
-

**Comment by bs7301** [comment]
최종결과가 어떻게 됐을까요? 궁금합니다 ..!
